### DISCUSSION TITLE
Proposal: Auto implicit casting between interfaces

### DISCUSSION THREAD

**Discussion by hez2010** [dotnet/csharplang#3340] [Language Ideas]
### Proposal
Consider below interfaces which are from different assemblies:
```csharp
namespace Foo
{
    interface IActionA
    {
        Baz Result { get; set; }
        void Do();
    }
}

namespace Bar
{
    interface IActionB
    {
        Baz Result { get; set; }
        void Do();
    }
}
```

They are absolutely same interfaces so an implicit casting between `Foo.IActionA` and `Bar.IActionB` can be added automatically so that a parameter accepts a `Foo.IActionA` can also accept a `Bar.IActionB`.

Interfaces which has same properties and methods can be defined as 'same'.

### Usage scenario
1. There're many interfaces which are same between namespace `Windows.Foundation` and `System`. With auto implicit casting, an object from `System` can be easily pass to methods which defined in `Windows.Foundation` without unnecessary casting.

2. If we want to use an interface which defined in another assembly (called `IA`), we will no longer need to reference the assembly, and we can define a interface (called `IB`) which is same as `IA` instead. And if there's some breaking changes happened in `IA`, the auto implicit casting will also be disallowed between `IA` and `IB`.

**Comment by YairHalberstadt** [comment]
This is an example of structural typing. C# has always been a nominally typed language and changing that would be a major change in philosophy/paradigm for the language.
It is possible that the some form of structural typing support will come as part of the shapes proposals that will hopefully be investigated for the C# 10 timeframe. See https://github.com/dotnet/csharplang/issues/1711 and #164.

> Default Interface Implementation
Since C# 8 introduced default Interface Implementation, same interfaces can behave differently.
Solution: disallow auto implicit conversion for interfaces which has default implementation.

This wouldn't work, as that would make adding a member with a Default Interface implementation a breaking change, which removes the entire point of default interface implementations!

**Comment by YairHalberstadt** [comment]
> Default Interface Implementation
Since C# 8 introduced default Interface Implementation, same interfaces can behave differently.

Also from a structural typing perspective I'm not sure this is an issue. You only care what methods the interface provides, not what they do. It's perfectly OK to override the DIM, so why shouldn't it be OK to give an instance of an interface with a different DIM?

One thing you don't address is how this would be implemented?

**Comment by hez2010** [comment]
> It's perfectly OK to override the DIM, so why shouldn't it be OK to give an instance of an interface with a different DIM?

I've removed the problem section in the post.

**Comment by hez2010** [comment]
> One thing you don't address is how this would be implemented?

While passing an object to a method which accepts an interface, compiler can check whether the object has properties and methods defined in target interface and cast it to the target interface.

**Comment by YairHalberstadt** [comment]
> and cast it to the target interface.

That doesn't really explain how it works. The runtime currently does not support structural typing, or casting from one interface to the other, and doing so would be a major change (I.e. adding generics in .Net 2.0 kind of change).

Can you give more details about how this would work?

One potential option is generating a wrapper object:

```csharp
class IActionAWrapperForAction2 : IActionA
{
    IAction2 _impl;
    public IActionAWrapperForAction2(IAction2 impl) => _impl = impl;
    Baz Result { get => _impl.Baz; set => _impl.Baz = value; }
    void Do() => _impl.Do();
}
```

And then a call is converted to `M(new IActionAWrapperForAction2(actionB));`

This has a number of disadvantages though:
1. Hidden allocation.
2. Loss of Reference Equality
3. Extra interface call (loss of performance)

These aren't critical, but they are relevant.

**Comment by hez2010** [comment]
@YairHalberstadt 
It works except when use reflection:

```csharp
class IActionAWrapperForAction2 : IActionA
{
    IAction2 _impl;
    public IActionAWrapperForAction2(IAction2 impl) => _impl = impl;
    Baz Result { get => _impl.Baz; set => _impl.Baz = value; }
    void Do() => _impl.Do();
}

IActionA x;
Foo(x);

void Foo(IAction2 obj)
{
    Assert.NotNull(obj.GetType().GetInterface("IAction2")); // boom!
}
```

Currently I can't figure out how to implement it without type erasure, but type erasure is a bad option in many cases.

**Comment by YairHalberstadt** [comment]
I don't see how type erasure would help here

**Comment by huoyaoyuan** [comment]
> They are absolutely same interfaces

C# is designed to treat them as *absolutely different* interfaces. The tutorial of Explicit Interface Method Implementation explains this.

I think *shape* is more suitable for this case.

**Comment by Shadowblitz16** [comment]
This would be good for things like INumber<T> which uses Create<TOther>() to create it.
And then we don't have to repeatedly call T.Create(0) to declare and array of generic numbers
